src/MPGDCylinderBarrelTracker_geo.cpp

```diff
@@ -0,0 +1,563 @@
+ *
+ * \ingroup trackers
+ *
+ * @{
+ */
+#include "DD4hep/DetFactoryHelper.h"
+#include "DD4hep/Printout.h"
+#include "DD4hep/Shapes.h"
+#include "DDRec/DetectorData.h"
+#include "DDRec/Surface.h"
+#include "XML/Layering.h"
+#include "XML/Utilities.h"
+#include <array>
+#include "DD4hepDetectorHelper.h"
+
+using namespace std;
+using namespace dd4hep;
+using namespace dd4hep::rec;
+using namespace dd4hep::detail;
+
+#include "Math/Vector2D.h"
+using namespace ROOT::Math;
+
+/** Micromegas Barrel Tracker with space frame
+ *
+ * - Designed to process "mpgd_barrel.xml" ("mpgd_barrel_ver1" as of 2024/02).
+ *
+ * - Derived from "BarrelTrackerWithFrame_geo.cpp".
+ *
+ * - "support" tag not addressed.
+ *
+ * - "frame" tag within the module element.
+ *
+ * - Several models of modules, with each a distinct radius of curvature
+ *  but a single XML <module> and a single <layer>.
+ *
+ * \code
+ * \endcode
+ *
+ *
+ * @author Yann Bedfer
+ */
+static Ref_t create_MPGDCylinderBarrelTracker(Detector& description, xml_h e,
+                                              SensitiveDetector sens) {
+  xml_det_t x_det = e;
+  Material air    = description.air();
+  int det_id      = x_det.id();
+  string det_name = x_det.nameStr();
+  DetElement sdet(det_name, det_id);
+
+  vector<Volume> volumes;
+  vector<PlacedVolume> sensitives;
+  vector<VolPlane> volplane_surfaces;
+
+  PlacedVolume pv;
+
+  //#define DEBUG_MPGDCylinderBarrelTracker
+#ifdef DEBUG_MPGDCylinderBarrelTracker
+  // TEMPORARILY INCREASE VERBOSITY level for debugging purposes
+  PrintLevel priorPrintLevel = printLevel();
+  setPrintLevel(DEBUG);
+#endif
+
+  // Set detector type flag
+  dd4hep::xml::setDetectorTypeFlag(x_det, sdet);
+  auto& params = DD4hepDetectorHelper::ensureExtension<dd4hep::rec::VariantParameters>(sdet);
+
+  // Add the volume boundary material if configured
+  for (xml_coll_t bmat(x_det, _Unicode(boundary_material)); bmat; ++bmat) {
+    xml_comp_t x_boundary_material = bmat;
+    DD4hepDetectorHelper::xmlToProtoSurfaceMaterial(x_boundary_material, params,
+                                                    "boundary_material");
+  }
+
+  dd4hep::xml::Dimension dimensions(x_det.dimensions());
+  Assembly assembly(det_name);
+
+  sens.setType("tracker");
+
+  // ********** MODULE
+  // ***** ONE AND ONLY ONE MODULE
+  xml_coll_t modules(x_det, _U(module));
+  if (modules.size() != 1) {
+    // Present detector constructor can only handle ONE <module> tag
+    printout(ERROR, "MPGDCylinderBarrelTracker", "Number of modules = %d. Must be = 1",
+             (int)modules.size());
+    throw runtime_error("Logics error in building modules.");
+  }
+  xml_comp_t x_mod   = modules;
+  string m_nam       = x_mod.nameStr();
+  double stave_width = dimensions.width(), det_length = dimensions.length();
+  printout(DEBUG, "MPGDCylinderBarrelTracker", "Module \"%s\" width = %.2f cm length = %.2f cm",
+           m_nam.c_str(), stave_width, det_length);
```

Comment:
Generally safer to divide out units when printing value with explicit unit:
```suggestion
  printout(DEBUG, "MPGDCylinderBarrelTracker", "Module \"%s\" width = %.2f cm length = %.2f cm",
           m_nam.c_str(), stave_width / cm, det_length / cm);
```

Reply:
Applies below.